### PR TITLE
nixosTest: make modular

### PR DESCRIPTION
**Pull request by roberth**
###### Why

 - Generated docs
 - Type checking
 - Overrides for interactive use
 - Reusable test fixtures with `imports`
 - Replace test logic with `disabledModules`
 - Work towards generalized test matrix feature (#176557)
 - Some other fixes

Closes #171396

##### What

 - Move `testing-python.nix` logic and `build-vms.nix` logic into modules
 - Rewrite the node numbering logic
 - Write docs

Extracted from #176557 

Hydra: 4299c1749d35ab472d56174404ba420df9d3febe https://hydra.nixos.org/eval/1781169?compare=1780016

First nixos-unstable: https://hydra.nixos.org/eval/1782482

"removed" attributes
 - `ec2`: `meta.broken` now applies to test runner
 - `fcitx`: `meta.broken` now applies to test runner
 - `gocd-agent`: `meta.broken` now applies to test runner
 - `hadoop`: rename the attr path to include system
 - `hocker-fetchdocker`: `meta.broken` now applies to test runner
 - `installed-tests.xdg-desktop-portal`: `meta.broken` now applies to test runner
 - `mariadb`, `mysql`: now runs on both `system` types, changing the attr path
 - `pulseaudio`: now runs on both `system` types, changing the attr path
 - `shadowsocks`: now runs on both `system` types, changing the attr path
 - `zigbee2mqtt`: `name` has already been added in master. merge will fix this. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
Ofborg ran out of disk space.

@ofborg eval

**Comment by roberth** [comment]
Rebased. Already reviewed as part of https://github.com/NixOS/nixpkgs/pull/176557. Will merge in a couple of days to avoid more rebasing.

File: nixos/doc/manual/development/writing-nixos-tests.section.md [hunk 1]
```diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
...
+Outside the `nixpkgs` repository, you can instantiate the test by first importing the NixOS library,
+
+```nix
+# regular nix
+let nixos-lib = import (nixpkgs + "/nixos/lib") { };
+in
+```
+
+```nix
+# flake
```
Comment (fricklerhandwerk) [inline nixos/doc/manual/development/writing-nixos-tests.section.md:79]: The Nixpkgs manual does not refer to flakes at all, and I wonder if we should keep it consistent here as well.

I'm inclined to await a community wide call to migrate to flakes and until then do not add scattered bits.

File: nixos/doc/manual/development/writing-nixos-tests.section.md [hunk 2]
```diff
@@ -121,7 +182,7 @@ The following methods are available on machine objects:
     least one will be returned.
 
     ::: {.note}
-    This requires passing `enableOCR` to the test attribute set.
+    This requires [`enableOCR`](#opt-enableOCR) to be set to `true`.
```
Comment (fricklerhandwerk) [inline nixos/doc/manual/development/writing-nixos-tests.section.md:185]: This seems unrelated. It's correct though so I don't mind.

**Review by fricklerhandwerk — APPROVED** [review]
LGTM wrt. docs. Only a few notes. (Did not inspect the code in any depth!)

Thank you so much for your work and patience with reviews.

**Comment by Atemu** [comment]
This broke kernel passthru tests:

```
$ nix-instantiate -A linux.passthru.tests --system x86_64-linux --show-trace
error: attempt to call something which is not a function but a set

       at /Users/atemu/Projects/nixpkgs/pkgs/os-specific/linux/kernel/generic.nix:217:11:

          216|       };
          217|     in [ (nixosTests.kernel-generic.testsForKernel overridableKernel) ] ++ kernelTests;
             |           ^
          218|   };
```

**Comment by roberth** [comment]
The way both `nixosTests` and `nixos/release.nix` load and transform the tests is incredibly ad hoc and fragile.
It appears that `nixosTests` now tries to invoke the test according what we might call the standalone `import ./make-test-python.nix` pattern, considering that `testsForLinuxPackages` appears to receive `{ pkgs = ...; system = ...; }` as its argument. Both `discoverTests` and `handleTest` pass these arguments, but perhaps `discoverTests` is at fault. I really can't tell though, because of all the "legacy" that this code needs to support.

Ideally we'd standardize the solution for both the test matrix "problem" and the package testing problem in one go. That could be https://github.com/NixOS/nixpkgs/pull/176557 or something similar.
